### ISSUE TITLE
Error creating release upgrade: Bad instruction: {update,'Elixir.Phoenix.PubSub.PG2',supervisor,[]}

### ISSUE THREAD

**Issue by smpallen99** [bitwalker/distillery#141]
I have project where:
* Creating initial release on master
* Created branch and did some changes
* Created upgrade version 0.0.2 on branch
* More dev on the branch
* Created upgrade version 0.1.0
* Everything worked so far
* Check out master
* Merge branch
* Create upgrade with `./node_modules/brunch/bin/brunch b -p && MIX_ENV=prod mix do compile, phoenix.digest, release --env=prod --upgrade --verbose`

I get the error below. 

```
[root@localhost ucx_remote]# ./node_modules/brunch/bin/brunch b -p && MIX_ENV=prod mix do compile, phoenix.digest, release --env=prod --upgrade --verbose
23 Nov 16:49:02 - info: compiling
23 Nov 16:49:02 - info: compiled 11 files into 7 files, copied 22 in 6.6 sec
Check your digested files at "priv/static"
==> Loading configuration..
==> Assembling release..
==> Building release ucx_remote:0.1.0 using environment prod
==> Discovered applications:
  trailing_format_plug-0.0.4
    from: _build/prod/lib/trailing_format_plug
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :cowboy
    includes: none

  neotoma-1.7.3
    from: _build/prod/lib/neotoma
    applications:
      :kernel
      :stdlib
    includes: none

  conform-2.1.1
    from: _build/prod/lib/conform
    applications:
      :kernel
      :stdlib
      :elixir
      :neotoma
    includes: none

  exactor-2.2.2
    from: _build/prod/lib/exactor
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

  canada-1.0.1
    from: _build/prod/lib/canada
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  utils-0.1.0
    from: _build/prod/lib/utils
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

  logger_file_backend-0.0.9-dev
    from: _build/prod/lib/logger_file_backend
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  gen_smtp-0.11.0
    from: _build/prod/lib/gen_smtp
    applications:
      :kernel
      :stdlib
      :crypto
      :asn1
      :public_key
      :ssl
    includes: none

  xain-0.6.0
    from: _build/prod/lib/xain
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

  ex_queb-0.2.2
    from: _build/prod/lib/ex_queb
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

  csvlixir-1.0.0
    from: _build/prod/lib/csvlixir
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  scrivener_ecto-1.0.3
    from: _build/prod/lib/scrivener_ecto
    applications:
      :kernel
      :stdlib
      :elixir
      :scrivener
      :logger
    includes: none

  scrivener-2.1.1
    from: _build/prod/lib/scrivener
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

  inflex-1.8.0
    from: _build/prod/lib/inflex
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  ex_admin-0.8.3-dev
    from: _build/prod/lib/ex_admin
    applications:
      :kernel
      :stdlib
      :elixir
      :gettext
      :phoenix
      :ecto
      :inflex
      :scrivener
      :scrivener_ecto
      :csvlixir
      :logger
      :ex_queb
      :xain
    includes: none

  crutches-1.0.0
    from: _build/prod/lib/crutches
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

  canary-1.0.0
    from: _build/prod/lib/canary
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

  ssh-4.2.2
    from: /usr/lib64/erlang/lib/ssh-4.2.2
    applications:
      :kernel
      :stdlib
      :crypto
      :public_key
    includes: none

  sshex-2.1.0
    from: _build/prod/lib/sshex
    applications:
      :kernel
      :stdlib
      :elixir
      :ssh
    includes: none

  timex_ecto-1.1.3
    from: _build/prod/lib/timex_ecto
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :ecto
      :timex
    includes: none

  combine-0.9.2
    from: _build/prod/lib/combine
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  tzdata-0.5.9
    from: _build/prod/lib/tzdata
    applications:
      :kernel
      :stdlib
      :elixir
      :hackney
      :logger
    includes: none

  timex-2.2.1
    from: _build/prod/lib/timex
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :tzdata
      :gettext
      :combine
    includes: none

  phoenix_timex-1.0.1
    from: _build/prod/lib/phoenix_timex
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :timex
    includes: none

  metrics-1.0.1
    from: _build/prod/lib/metrics
    applications:
      :kernel
      :stdlib
    includes: none

  ssl_verify_fun-1.1.0
    from: _build/prod/lib/ssl_verify_fun
    applications:
      :kernel
      :stdlib
      :ssl
    includes: none

  certifi-0.4.0
    from: _build/prod/lib/certifi
    applications:
      :kernel
      :stdlib
    includes: none

  mimerl-1.0.2
    from: _build/prod/lib/mimerl
    applications:
      :kernel
      :stdlib
    includes: none

  idna-1.2.0
    from: _build/prod/lib/idna
    applications:
      :kernel
      :stdlib
    includes: none

  ssl-7.3
    from: /usr/lib64/erlang/lib/ssl-7.3
    applications:
      :crypto
      :public_key
      :kernel
      :stdlib
    includes: none

  public_key-1.1.1
    from: /usr/lib64/erlang/lib/public_key-1.1.1
    applications:
      :asn1
      :crypto
      :kernel
      :stdlib
    includes: none

  asn1-4.0.2
    from: /usr/lib64/erlang/lib/asn1-4.0.2
    applications:
      :kernel
      :stdlib
    includes: none

  hackney-1.6.1
    from: _build/prod/lib/hackney
    applications:
      :kernel
      :stdlib
      :crypto
      :asn1
      :public_key
      :ssl
      :idna
      :mimerl
      :certifi
      :ssl_verify_fun
      :metrics
    includes: none

  httpoison-0.9.0
    from: _build/prod/lib/httpoison
    applications:
      :kernel
      :stdlib
      :elixir
      :hackney
    includes: none

  swoosh-0.4.0
    from: _build/prod/lib/swoosh
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :poison
      :httpoison
    includes: none

  phoenix_swoosh-0.1.3
    from: _build/prod/lib/phoenix_swoosh
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :swoosh
    includes: none

  uuid-1.1.4
    from: _build/prod/lib/uuid
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  comeonin-2.5.2
    from: _build/prod/lib/comeonin
    applications:
      :kernel
      :stdlib
      :elixir
      :crypto
      :logger
    includes: none

  coherence-0.3.0
    from: _build/prod/lib/coherence
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :comeonin
      :ecto
      :uuid
      :phoenix_swoosh
      :phoenix_timex
      :timex_ecto
      :tzdata
    includes: none

  connection-1.0.4
    from: _build/prod/lib/connection
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  db_connection-1.0.0
    from: _build/prod/lib/db_connection
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :connection
    includes: none

  mariaex-0.7.9
    from: _build/prod/lib/mariaex
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :crypto
      :decimal
      :db_connection
    includes: none

  poolboy-1.5.1
    from: _build/prod/lib/poolboy
    applications:
      :kernel
      :stdlib
    includes: none

  decimal-1.3.1
    from: _build/prod/lib/decimal
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  ecto-2.0.5
    from: _build/prod/lib/ecto
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :decimal
      :poolboy
    includes: none

  phoenix_ecto-3.0.1
    from: _build/prod/lib/phoenix_ecto
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :ecto
      :plug
    includes: none

  gettext-0.12.1
    from: _build/prod/lib/gettext
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
    includes: none

  cowlib-1.0.2
    from: _build/prod/lib/cowlib
    applications:
      :kernel
      :stdlib
      :crypto
    includes: none

  ranch-1.2.1
    from: _build/prod/lib/ranch
    applications:
      :kernel
      :stdlib
    includes: none

  cowboy-1.0.4
    from: _build/prod/lib/cowboy
    applications:
      :kernel
      :stdlib
      :ranch
      :cowlib
      :crypto
    includes: none

  phoenix_html-2.8.0
    from: _build/prod/lib/phoenix_html
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :plug
    includes: none

  phoenix_pubsub-1.0.1
    from: _build/prod/lib/phoenix_pubsub
    applications:
      :kernel
      :stdlib
      :elixir
      :logger
      :crypto
    includes: none

  eex-1.3.4
    from: /usr/local/lib/elixir/bin/../lib/eex
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  poison-2.2.0
    from: _build/prod/lib/poison
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  mime-1.0.1
    from: _build/prod/lib/mime
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  logger-1.3.4
    from: /usr/local/lib/elixir/bin/../lib/logger
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  crypto-3.6.3
    from: /usr/lib64/erlang/lib/crypto-3.6.3
    applications:
      :kernel
      :stdlib
    includes: none

  plug-1.2.2
    from: _build/prod/lib/plug
    applications:
      :kernel
      :stdlib
      :elixir
      :crypto
      :logger
      :mime
    includes: none

  phoenix-1.2.1
    from: _build/prod/lib/phoenix
    applications:
      :kernel
      :stdlib
      :elixir
      :plug
      :poison
      :logger
      :eex
    includes: none

  compiler-6.0.3
    from: /usr/lib64/erlang/lib/compiler-6.0.3
    applications:
      :kernel
      :stdlib
    includes: none

  elixir-1.3.4
    from: /usr/local/lib/elixir/bin/../lib/elixir
    applications:
      :kernel
      :stdlib
      :compiler
    includes: none

  stdlib-2.8
    from: /usr/lib64/erlang/lib/stdlib-2.8
    applications:
      :kernel
    includes: none

  kernel-4.2
    from: /usr/lib64/erlang/lib/kernel-4.2
    applications: none
    includes: none

  ucx_remote-0.1.0
    from: _build/prod/lib/ucx_remote
    applications:
      :kernel
      :stdlib
      :elixir
      :phoenix
      :phoenix_pubsub
      :phoenix_html
      :cowboy
      :logger
      :gettext
      :phoenix_ecto
      :mariaex
      :coherence
      :sshex
      :canary
      :crutches
      :ex_admin
      :gen_smtp
      :logger_file_backend
      :utils
      :canada
      :exactor
      :conform
      :trailing_format_plug
    includes: none

  iex-1.3.4
    from: /usr/local/lib/elixir/bin/../lib/iex
    applications:
      :kernel
      :stdlib
      :elixir
    includes: none

  sasl-2.7
    from: /usr/lib64/erlang/lib/sasl-2.7
    applications:
      :kernel
      :stdlib
    includes: none

==> conform: loading schema
==> conform: generating escript
==> conform: done!
==> Generated overlay vars:
    release_name=:ucx_remote
    release_version="0.1.0"
    is_upgrade=true
    upgrade_from="0.0.2"
    dev_mode=false
    include_erts=true
    include_src=false
    include_system_libs=true
    erl_opts=""
    erts_vsn="7.3"
    output_dir="rel/ucx_remote"
==> Copying applications to rel/ucx_remote
==> Generating nodetool
==> Generating start_erl.data
==> Generating vm.args
==> Generating sys.config from config/config.exs
==> Including ERTS 7.3 from /usr/lib64/erlang/erts-7.3
==> Generating boot script
==> Generating RELEASES
==> Generating start_clean.boot
==> Generating relup for ucx_remote
==> db_connection requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for db_connection 1.0.0-rc.5 -> 1.0.0
==> decimal requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for decimal 1.1.2 -> 1.3.1
==> ecto requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for ecto 2.0.4 -> 2.0.5
==> ex_admin requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for ex_admin 0.8.2 -> 0.8.3-dev
==> exactor requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for exactor 2.2.1 -> 2.2.2
==> gettext requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for gettext 0.11.0 -> 0.12.1
==> inflex requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for inflex 1.7.0 -> 1.8.0
==> mariaex requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for mariaex 0.7.7 -> 0.7.9
==> phoenix_html requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for phoenix_html 2.6.2 -> 2.8.0
==> phoenix_pubsub requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for phoenix_pubsub 1.0.0 -> 1.0.1
==> plug requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for plug 1.2.0 -> 1.2.2
==> scrivener_ecto requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for scrivener_ecto 1.0.2 -> 1.0.3
==> ucx_remote requires an appup, but it wasn't provided, one will be generated for you..
==> Generated .appup for ucx_remote 0.0.2 -> 0.1.0
==> Failed to build release:

    Bad instruction: {update,'Elixir.Phoenix.PubSub.PG2',supervisor,[]}
```

**Comment by kennyballou** [comment]
I have a suspicion this is easier to reproduce when creating a new release of an application that has supervisor changes, specifically changes to the supervisor tree and its children processes. I see the similar output when attempting to create a release between the [v0.1.0][2] tag and [v0.2.0][3] tag of this [demo][1] application.

[1]: https://github.com/kennyballou/octochat
[2]: https://github.com/kennyballou/octochat/releases/tag/v0.1.0
[3]: https://github.com/kennyballou/octochat/releases/tag/v0.2.0

**Comment by bitwalker** [comment]
I'll take a look, I suspect that instruction is not valid for supervisors, and a different one is needed.

**Comment by kennyballou** [comment]
It looks [valid][1] to me, but I'm not sure. I haven't tried this manually yet.

[1]: http://erlang.org/doc/design_principles/appup_cookbook.html#id85798

**Comment by talklittle** [comment]
Distillery is adding a 4th tuple element `[]` for `:update`, not in the cookbook that @kennyballou linked. Is that the bug?

If so then https://github.com/bitwalker/distillery/blob/41e2f8abf53001089aecc6268cea79ef4bc30e93/test/appup_test.exs probably needs to be updated too.

It differs from Exrm's `:update` tuple format too: https://github.com/bitwalker/exrm/blob/417286f13aebb4a19bee39bfdb4cd44b41d18f5f/test/appups_test.exs

**Comment by kennyballou** [comment]
Yeah, I missed that pattern, I thought the extra `[]` was valid since a supervisor _could_ have dependent modules(?).

**Comment by bitwalker** [comment]
This is addressed via #150, so I'm closing for now.
